WcaOnRails/app/controllers/application_controller.rb

```diff
@@ -14,8 +14,32 @@ class ApplicationController < ActionController::Base
   end
 
   def set_locale
-    # We set this to have access to the locale in controller
-    I18n.locale = params[:locale] || I18n.default_locale
+    # If the locale for the session is not set, we want to infer it from the following sources:
+    #  - the current user preferred locale
+    #  - the Accept-Language http header
+    unless session[:locale]
+      session[:locale] = if current_user
+                           current_user.preferred_locale
+                         else
+                           http_accept_language.preferred_language_from(I18n.available_locales)
+                         end
```

Comment:
Not sure if that's more sweet but it can also be this way
```ruby
session[:locale] = current_user&.preferred_locale || http_accept_language.preferred_language_from(I18n.available_locales)
```